File: android/src/main/kotlin/dev/nikkothe/epson_pos_printer/printer/AsyncCallbackDelegate.kt
package dev.nikkothe.epson_pos_printer.printer

import com.epson.epos2.printer.Printer
import com.epson.epos2.printer.PrinterStatusInfo
import com.epson.epos2.printer.ReceiveListener

class PrinterAsyncDelegate(
    private val callback: FlutterAsyncCallback,
    private val printer: Printer
) : ReceiveListener {

    init {
        printer.setReceiveEventListener(this)
    }

    override fun onPtrReceive(printerObj: Printer?, code: Int, status: PrinterStatusInfo?, printJobId: String?) {
        // Callback should be only called once, it should be cleared after the use
        printer.setReceiveEventListener(null)

        try {
            val printerStatus = mapOf(
                "printerJobId" to printJobId,
                "connection" to (status?.connection == 1),
                "online" to (status?.online == 1),
                "coverOpen" to (status?.coverOpen == 1),
                "paper" to status?.paper,
                "paperFeed" to (status?.paperFeed == 0),
                "panelSwitch" to (status?.panelSwitch == 0),
                "waitOnline" to (status?.waitOnline?.toInt() ?: 0),
                "drawer" to (status?.drawer == 0),
                "errorStatus" to status?.errorStatus,
                "autoRecoverError" to status?.autoRecoverError,
                "buzzer" to (status?.buzzer == 0),
                "adapter" to (status?.adapter == 0),
                "batteryLevel" to status?.batteryLevel,
                "removalWaiting" to status?.removalWaiting,
                "unrecoverError" to status?.unrecoverError
            )

            checkCallbackCode(code, printerStatus)

            callback(printerStatus)
        } catch (e: Exception) {
            throw e
        }
    }
}

File: ios/Classes/Printer/AsyncCallbackDelegate.swift
import Foundation

class PrinterAsyncDelegate :  NSObject, Epos2PtrReceiveDelegate  {
    unowned private let printer: Epos2Printer
    private let callback: FlutterAsyncCallback

    init(_ callback: @escaping FlutterAsyncCallback, attachTo printer: Epos2Printer) {
        self.printer = printer
        self.callback = callback
        
        super.init()
        
        printer.setReceiveEventDelegate(self)
    }

    func onPtrReceive(_ printerObj: Epos2Printer!, code: Int32, status: Epos2PrinterStatusInfo, printJobId: String!) {
        // Callback should be only called once, it should be cleared after the use
        printer.setReceiveEventDelegate(nil)    
    
        do {
            let printerStatus: PrinterStatus = [
                "printerJobId": printJobId,
                "connection": (status.connection == 1),
                "online": (status.online == 1),
                "coverOpen": (status.coverOpen == 1),
                "paper": status.paper,
                "paperFeed": (status.paperFeed == 0),
                "panelSwitch": (status.panelSwitch == 0),
                "waitOnline": (status.waitOnline),
                "drawer": (status.drawer == 0),
                "errorStatus": status.errorStatus,
                "autoRecoverError": status.autoRecoverError,
                "buzzer": (status.buzzer == 0),
                "adapter": (status.adapter == 0),
                "batteryLevel": status.batteryLevel,
                "removalWaiting": status.removalWaiting,
                "unrecoverError": status.unrecoverError
            ]
                        
            try check(callbackCode: code, printerStatus: printerStatus, methodOverride: "Error with status")
            
            callback(printerStatus)
        } catch {
            print("Error onPtrReceive \(error)")
            callback(flutterError(fromError: error, method: "Epos2Printer.onPtrReceive"))
        }
    }
}
